### PR TITLE
Implement ipaddressallocation controller

### PR DESCRIPTION
**Pull request by zhengxiexie**
IPAddressAllocation with Private ip_address_block_visibility can be realized successfully.
```
apiVersion: nsx.vmware.com/v1alpha1
kind: IPAddressAllocation
metadata:
  name: guestcluster-workers-a
  namespace: zx-test
spec:
  ip_address_block_visibility: Private
  allocation_size: 32
```
image

IPAddressAllocation with External ip_address_block_visibility can be realized successfully

```
apiVersion: nsx.vmware.com/v1alpha1
kind: IPAddressAllocation
metadata:
  name: guestcluster-workers-b
  namespace: zx-test
spec:
  ip_address_block_visibility: External
  allocation_size: 32
```
image

The default ip_address_block_visibility for IPAddressAllocation will be Private if ip_address_block_visibility is not defined.
```
apiVersion: nsx.vmware.com/v1alpha1
kind: IPAddressAllocation
metadata:
  name: guestcluster-workers-b
  namespace: zx-test
spec:
  allocation_size: 32
```
image

Both could be deleted successfully.

If the first ipblcok is exhausted, will try next one
```
apiVersion: nsx.vmware.com/v1alpha1
kind: IPAddressAllocation
metadata:
  name: guestcluster-workers-c
  namespace: zx-test-2
spec:
  ip_address_block_visibility: Private
  allocation_size: 131072
```
image

image

If all the blocks are exhausted, will return error
```
apiVersion: nsx.vmware.com/v1alpha1
kind: IPAddressAllocation
metadata:
  name: guestcluster-workers-d
  namespace: zx-test-2
spec:
  ip_address_block_visibility: Private
  allocation_size: 33554432
```
image

### CODE REVIEW

File: pkg/nsx/services/ipaddressallocation/ipaddressallocation.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+		return false, err
+	}
+	existingIPAddressAllocation, err := service.indexedIPAddressAllocation(obj.UID)
+	if err != nil {
+		log.Error(err, "failed to get ipaddressallocation", "UID", obj.UID)
+		return false, err
+	}
+	log.V(1).Info("existing ipaddressallocation", "ipaddressallocation", existingIPAddressAllocation)
+	_ = common.CompareResource(IpAddressAllocationToComparable(existingIPAddressAllocation),
+		IpAddressAllocationToComparable(nsxIPAddressAllocation))
```
Comment (heypnus) [inline pkg/nsx/services/ipaddressallocation/ipaddressallocation.go:75]: It seems that we compared them but do nothing?

File: pkg/nsx/services/ipaddressallocation/ipaddressallocation.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+	if len(VPCInfo) == 0 {
+		err = util.NoEffectiveOption{Desc: "no valid org and project for ipaddressallocation"}
+		return err
+	}
+	err = service.NSXClient.IPAddressAllocationClient.Patch(VPCInfo[0].OrgID, VPCInfo[0].ProjectID, VPCInfo[0].ID, *nsxIPAddressAllocation.Id, *nsxIPAddressAllocation)
+	err = util.NSXApiError(err)
+	if err != nil {
+		// not return err, try to get it from nsx, in case if cidr not realized at the first time
+		// support realize multiple times
+		log.Error(err, "patch failed, try to get it from nsx")
```
Comment (heypnus) [inline pkg/nsx/services/ipaddressallocation/ipaddressallocation.go:111]: Please add the object info for troubleshooting

File: pkg/nsx/services/ipaddressallocation/ipaddressallocation.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+	VPCInfo := service.VPCService.ListVPCInfo(ns)
+	if len(VPCInfo) == 0 {
+		err = util.NoEffectiveOption{Desc: "no valid org and project for ipaddressallocation"}
+		return err
+	}
+	err = service.NSXClient.IPAddressAllocationClient.Patch(VPCInfo[0].OrgID, VPCInfo[0].ProjectID, VPCInfo[0].ID, *nsxIPAddressAllocation.Id, *nsxIPAddressAllocation)
+	err = util.NSXApiError(err)
+	if err != nil {
+		// not return err, try to get it from nsx, in case if cidr not realized at the first time
+		// support realize multiple times
```
Comment (heypnus) [inline pkg/nsx/services/ipaddressallocation/ipaddressallocation.go:110]: `support realize multiple times` -- is it a TODO or you already implemented it in the current patch set?

File: pkg/nsx/services/ipaddressallocation/ipaddressallocation.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+			return err
+		}
+	case types.UID:
+		nsxIPAddressAllocation, err = service.indexedIPAddressAllocation(o)
+		if err != nil {
+			log.Error(err, "failed to get ipaddressallocation by UID", "UID", o)
+			return err
+		}
+	}
+	if nsxIPAddressAllocation == nil {
```
Comment (heypnus) [inline pkg/nsx/services/ipaddressallocation/ipaddressallocation.go:148]: Add a log to indicate the object not found before returning nil?

File: cmd/main.go [hunk 5]
```diff
@@ -147,6 +149,21 @@ func StartNamespaceController(mgr ctrl.Manager, cf *config.NSXOperatorConfig, vp
...
+	ipAddressAllocationReconciler := &ipaddressallocation.IPAddressAllocationReconciler{
+		Client:     mgr.GetClient(),
+		Scheme:     mgr.GetScheme(),
+		Service:    ipAddressAllocationService,
+		VPCService: vpcService,
+		Recorder:   mgr.GetEventRecorderFor("ipaddressallocation-controller"),
+	}
+
+	if err := ipAddressAllocationReconciler.SetupWithManager(mgr); err != nil {
+		log.Error(err, "failed to create namespace controller", "controller", "Namespace")
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply cmd/main.go:162]: ok

File: pkg/apis/v1alpha1/ipaddressallocation_types.go [hunk 6]
```diff
@@ -10,9 +10,8 @@ import (
 type IPAddressVisibility string
 
 const (
-	IPAddressVisibilityExternal = "External"
-	IPAddressVisibilityPrivate  = "Private"
-	IPAddressVisibilityProject  = "Project"
+	IPAddressVisibilityExternal = "EXTERNAL"
+	IPAddressVisibilityPrivate  = "PRIVATE"
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/apis/v1alpha1/ipaddressallocation_types.go:14]: deleted

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+type IPAddressAllocationReconciler struct {
+	client.Client
+	Scheme     *apimachineryruntime.Scheme
+	Service    *ipaddressallocation.IPAddressAllocationService
+	VPCService servicecommon.VPCServiceProvider
+	Recorder   record.EventRecorder
+}
+
+func deleteSuccess(r *IPAddressAllocationReconciler, _ *context.Context, o *v1alpha1.IPAddressAllocation) {
+	r.Recorder.Event(o, v1.EventTypeNormal, common.ReasonSuccessfulDelete, "IPPool CR has been successfully deleted")
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go:50]: ok

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+	servicecommon "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/ipaddressallocation"
+)
+
+var (
+	log           = logger.Log
+	once          sync.Once
+	resultNormal  = common.ResultNormal
+	resultRequeue = common.ResultRequeue
+	MetricResType = common.MetricResTypeIPPool
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go:37]: ok

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+			updateFail(r, &ctx, obj, &err)
+			return resultRequeue, err
+		}
+		log.V(1).Info("added finalizer on ipaddressallocation CR", "ipaddressallocation", req.NamespacedName)
+	}
+
+	updated, err := r.Service.CreateOrUpdateIPAddressAllocation(obj)
+	if err != nil {
+		updateFail(r, &ctx, obj, &err)
+		return resultRequeue, nil
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go:139]: ok

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+		CRIPAddressAllocationSet.Insert(string(ipa.UID))
+	}
+
+	log.V(2).Info("ipaddressallocation garbage collector", "nsxIPAddressAllocationSet", ipAddressAllocationSet, "CRIPAddressAllocationSet", CRIPAddressAllocationSet)
+
+	diffSet := ipAddressAllocationSet.Difference(CRIPAddressAllocationSet)
+	for elem := range diffSet {
+		log.Info("GC collected ipaddressallocation CR", "UID", elem)
+		if err := r.Service.DeleteIPAddressAllocation(types.UID(elem)); err != nil {
+			log.Error(err, "failed to delete ipaddressallocation CR", "UID", elem)
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go:214]: ok

File: pkg/nsx/services/ipaddressallocation/ipaddressallocation.go (hunk 4 above)
Comment (zhengxiexie) [reply pkg/nsx/services/ipaddressallocation/ipaddressallocation.go:148]: ok

File: pkg/nsx/services/ipaddressallocation/ipaddressallocation.go (hunk 3 above)
Comment (zhengxiexie) [reply pkg/nsx/services/ipaddressallocation/ipaddressallocation.go:110]: current patch, update the comment for misunderstanding

File: pkg/nsx/services/ipaddressallocation/ipaddressallocation.go (hunk 2 above)
Comment (zhengxiexie) [reply pkg/nsx/services/ipaddressallocation/ipaddressallocation.go:111]: ok

File: pkg/nsx/services/ipaddressallocation/ipaddressallocation.go (hunk 1 above)
Comment (zhengxiexie) [reply pkg/nsx/services/ipaddressallocation/ipaddressallocation.go:75]: oh, for test, forget to change it back

File: pkg/nsx/services/ipaddressallocation/ipaddressallocation.go [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+
+	wg.Add(1)
+
+	ipAddressAllocationService := &IPAddressAllocationService{Service: service, VPCService: vpcService}
+	ipAddressAllocationService.ipAddressAllocationStore = &IPAddressAllocationStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeIPAddressAllocationCRUID: indexFunc}),
+		BindingType: model.IpAddressPoolBindingType(),
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/nsx/services/ipaddressallocation/ipaddressallocation.go:41]: ok, since the query API not working, not hitting it, waiting...

File: build/yaml/crd/nsx.vmware.com_ipaddressallocations.yaml [hunk 12]
```diff
@@ -54,12 +54,10 @@ spec:
               ip_address_block_visibility:
                 default: Private
                 description: IPAddressBlockVisibility specifies the visibility of
-                  the IPBlocks to allocate IP addresses. Can be External, Private
-                  or Project.
+                  the IPBlocks to allocate IP addresses. Can be External or Private.
```
Comment (dantingl) [inline build/yaml/crd/nsx.vmware.com_ipaddressallocations.yaml:57]: Why don't we support Project?

File: build/yaml/crd/nsx.vmware.com_nsxserviceaccounts.yaml [hunk 13]
```diff
@@ -21,14 +20,19 @@ spec:
         description: NSXServiceAccount is the Schema for the nsxserviceaccounts API
         properties:
           apiVersion:
-            description: 'APIVersion defines the versioned schema of this representation
-              of an object. Servers should convert recognized schemas to the latest
-              internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources'
+            description: |-
```
Comment (dantingl) [inline build/yaml/crd/nsx.vmware.com_nsxserviceaccounts.yaml:23]: better to put all these description changes into a separate patch

File: cmd/main.go [hunk 14]
```diff
@@ -147,6 +149,21 @@ func StartNamespaceController(mgr ctrl.Manager, cf *config.NSXOperatorConfig, vp
...
+	ipAddressAllocationReconciler := &ipaddressallocation.IPAddressAllocationReconciler{
+		Client:     mgr.GetClient(),
+		Scheme:     mgr.GetScheme(),
+		Service:    ipAddressAllocationService,
+		VPCService: vpcService,
+		Recorder:   mgr.GetEventRecorderFor("ipaddressallocation-controller"),
+	}
+
+	if err := ipAddressAllocationReconciler.SetupWithManager(mgr); err != nil {
+		log.Error(err, "failed to create ipaddressallocation controller", "controller", "Namespace")
```
Comment (dantingl) [inline cmd/main.go:162]: remove `"controller", "Namespace"`?

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go [hunk 15]
```diff
@@ -0,0 +1,217 @@
+/* Copyright © 2023 VMware, Inc. All Rights Reserved.
```
Comment (dantingl) [inline pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go:1]: 2024

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+	r.Recorder.Event(o, v1.EventTypeWarning, common.ReasonFailUpdate, fmt.Sprintf("%v", *e))
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateFailTotal, MetricResType)
+}
+
+func (r *IPAddressAllocationReconciler) setReadyStatusFalse(ctx *context.Context, ipaddressallocation *v1alpha1.IPAddressAllocation, transitionTime metav1.Time, err *error) {
+	conditions := []v1alpha1.Condition{
+		{
+			Type:    v1alpha1.Ready,
+			Status:  v1.ConditionFalse,
+			Message: "NSX IPAddressAllocation could not be created/updated/deleted",
```
Comment (dantingl) [inline pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go:77]: `created/updated/deleted`: this message is not accurate about the operation

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+	e := r.Client.Status().Update(*ctx, ipaddressallocation)
+	if e != nil {
+		log.Error(e, "unable to update IPAddressAllocation status", "ipaddressallocation", ipaddressallocation)
+	}
+}
+
+func (r *IPAddressAllocationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	once.Do(func() { go common.GenericGarbageCollector(make(chan bool), servicecommon.GCInterval, r.collectGarbage) })
+	obj := &v1alpha1.IPAddressAllocation{}
+	log.Info("reconciling ipaddressallocation CR", "ipaddressallocation", req.NamespacedName)
```
Comment (dantingl) [inline pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go:112]: `ipaddressallocation` -> IPAddressAllocation
same for all the following messages

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+		updateSuccess(r, &ctx, obj)
+	}
+	return resultNormal, nil
+}
+
+func (r *IPAddressAllocationReconciler) handleDeletion(ctx context.Context, req ctrl.Request, obj *v1alpha1.IPAddressAllocation) (ctrl.Result, error) {
+	if controllerutil.ContainsFinalizer(obj, servicecommon.IPAddressAllocationFinalizerName) {
+		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, MetricResType)
+		if err := r.Service.DeleteIPAddressAllocation(obj); err != nil {
+			log.Error(err, "deletion failed, would retry exponentially", "ipaddressallocation", req.NamespacedName)
```
Comment (dantingl) [inline pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go:151]: ipaddressallocation -> IPAddressAllocation

File: pkg/controllers/nsxserviceaccount/nsxserviceaccount_controller.go [hunk 19]
```diff
@@ -215,7 +215,8 @@ func (r *NSXServiceAccountReconciler) validateRealized(count uint16, ca []byte,
 	return count, ca
 }
 
-func (r *NSXServiceAccountReconciler) garbageCollector(nsxServiceAccountUIDSet sets.Set[string], nsxServiceAccountList *nsxvmwarecomv1alpha1.NSXServiceAccountList) (gcSuccessCount, gcErrorCount uint32) {
+func (r *NSXServiceAccountReconciler) garbageCollector(nsxServiceAccountUIDSet sets.Set[string],
+	nsxServiceAccountList *nsxvmwarecomv1alpha1.NSXServiceAccountList) (gcSuccessCount, gcErrorCount uint32) {
```
Comment (dantingl) [inline pkg/controllers/nsxserviceaccount/nsxserviceaccount_controller.go:219]: remove this change

File: pkg/nsx/services/common/types.go [hunk 20]
```diff
@@ -53,6 +53,9 @@ const (
 	TagScopeIPPoolCRName               string = "nsx-op/ippool_name"
 	TagScopeIPPoolCRUID                string = "nsx-op/ippool_uid"
 	TagScopeIPPoolCRType               string = "nsx-op/ippool_type"
+	TagScopeIPAddressAllocationCRName  string = "nsx-op/ipaddressallocation_name"
+	TagScopeIPAddressAllocationCRUID   string = "nsx-op/ipaddressallocation_uid"
+	TagScopeIPAddressAllocationCRType  string = "nsx-op/ipaddressallocation_type"
```
Comment (dantingl) [inline pkg/nsx/services/common/types.go:58]: Remove type from tag, get it from CR

File: pkg/nsx/services/common/types.go [hunk 21]
```diff
@@ -84,14 +87,15 @@ const (
...
-	IPPoolFinalizerName = "ippool.nsx.vmware.com/finalizer"
-	DefaultSNATID       = "DEFAULT"
-	AVISubnetLBID       = "_AVI_SUBNET--LB"
-	IPPoolTypePublic    = "Public"
-	IPPoolTypePrivate   = "Private"
+	GCInterval                       = 60 * time.Second
+	RealizeTimeout                   = 2 * time.Minute
+	RealizeMaxRetries                = 3
+	IPPoolFinalizerName              = "ippool.nsx.vmware.com/finalizer"
+	IPAddressAllocationFinalizerName = "ipaddressallocation.nsx.vmware.com/finalizer"
```
Comment (dantingl) [inline pkg/nsx/services/common/types.go:94]: Put all FinalizerName together with line 100

File: pkg/nsx/services/ipaddressallocation/ipaddressallocation.go [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+		return false, nil
+	}
+
+	if err := service.Apply(nsxIPAddressAllocation); err != nil {
+		return false, err
+	}
+
+	createdIPAddressAllocation, err := service.indexedIPAddressAllocation(obj.UID)
+	if err != nil {
+		log.Error(err, "failed to get ipaddressallocation", "UID", obj.UID)
```
Comment (heypnus) [inline pkg/nsx/services/ipaddressallocation/ipaddressallocation.go:88]: better to make this log message different from L70, e.g. `"failed to get ipaddressallocation"` -> `"failed to get created ipaddressallocation"`

File: pkg/nsx/services/ipaddressallocation/store.go [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+func (ipAddressAllocationStore *IPAddressAllocationStore) Apply(i interface{}) error {
+	ipAddressAllocation := i.(*model.VpcIpAddressAllocation)
+	if ipAddressAllocation.MarkedForDelete != nil && *ipAddressAllocation.MarkedForDelete {
+		err := ipAddressAllocationStore.Delete(ipAddressAllocation)
+		log.V(1).Info("delete ipAddressAllocation from store", "ipAddressAllocation", ipAddressAllocation)
+		if err != nil {
+			return err
+		}
```
Comment (heypnus) [inline pkg/nsx/services/ipaddressallocation/store.go:56]:
```
		if err != nil {
			return err
		}
		log.V(1).Info("deleted ipAddressAllocation from store", "ipAddressAllocation", ipAddressAllocation)
```

File: pkg/nsx/services/ipaddressallocation/store.go [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+	} else {
+		err := ipAddressAllocationStore.Add(ipAddressAllocation)
+		log.V(1).Info("add ipAddressAllocation to store", "ipAddressAllocation", ipAddressAllocation)
+		if err != nil {
+			return err
+		}
```
Comment (heypnus) [inline pkg/nsx/services/ipaddressallocation/store.go:62]:
```
		if err != nil {
			return err
		}
		log.V(1).Info("added ipAddressAllocation to store", "ipAddressAllocation", ipAddressAllocation)
```

File: build/yaml/crd/nsx.vmware.com_ipaddressallocations.yaml (hunk 12 above)
Comment (zhengxiexie) [reply build/yaml/crd/nsx.vmware.com_ipaddressallocations.yaml:57]: I didn't find Project in nsx API doc.

File: cmd/main.go (hunk 14 above)
Comment (zhengxiexie) [reply cmd/main.go:162]: ok

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go (hunk 15 above)
Comment (zhengxiexie) [reply pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go:1]: ok

File: pkg/nsx/services/ipaddressallocation/store.go (hunk 24 above)
Comment (zhengxiexie) [reply pkg/nsx/services/ipaddressallocation/store.go:62]: ok

File: pkg/nsx/services/ipaddressallocation/store.go (hunk 23 above)
Comment (zhengxiexie) [reply pkg/nsx/services/ipaddressallocation/store.go:56]: ok

File: pkg/nsx/services/ipaddressallocation/ipaddressallocation.go (hunk 22 above)
Comment (zhengxiexie) [reply pkg/nsx/services/ipaddressallocation/ipaddressallocation.go:88]: ok

File: build/yaml/crd/nsx.vmware.com_ippools.yaml [hunk 25]
```diff
@@ -3,8 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.11.0
-  creationTimestamp: null
+    controller-gen.kubebuilder.io/version: v0.14.0
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply build/yaml/crd/nsx.vmware.com_ippools.yaml:6]: ok

File: pkg/nsx/services/common/types.go (hunk 21 above)
Comment (zhengxiexie) [reply pkg/nsx/services/common/types.go:94]: ok

File: pkg/nsx/services/common/types.go (hunk 20 above)
Comment (zhengxiexie) [reply pkg/nsx/services/common/types.go:58]: ok, not used any more

File: pkg/controllers/nsxserviceaccount/nsxserviceaccount_controller.go (hunk 19 above)
Comment (zhengxiexie) [reply pkg/controllers/nsxserviceaccount/nsxserviceaccount_controller.go:219]: ok

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go (hunk 18 above)
Comment (zhengxiexie) [reply pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go:151]: ok

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go (hunk 17 above)
Comment (zhengxiexie) [reply pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go:112]: ok, all modified.

File: cmd/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply cmd/main.go]: Now is `log.Error(err, "failed to create ipaddressallocation controller")`

**Review by dantingl — APPROVED** [review]

**Review by heypnus — APPROVED** [review]
